Common functions of libvmmapi encapsulated

sonicaj opened the pull request.

sonicaj committed:
Basic functionality complete

sonicaj committed:
types header file added

sonicaj committed:
Code refined

william-gr commented on bhyve.pyx:
```diff
@@ -1,4 +1,92 @@
+        self.vm_name = vm_name
+
+        if create_vm:
+            if bhyve.vm_create(vm_name.encode()) != 0:
```
Calling C functions (.e.g bhyve.*) should release the GIL (with nogil:). So these can run in a thread and not block other python code to run.

sonicaj committed:
Reviews Addressed (1)

sonicaj committed:
Reviews Addressed (2)

william-gr commented on bhyve.pyx:
```diff
@@ -1,4 +1,107 @@
+
+        return True
+
+    def get_stats(self, int vcpu=0):
```
How do I know how manu vcpus are there?

Perhaps we should return all of them?

william-gr commented on bhyve.pyx:
```diff
@@ -1,4 +1,107 @@
+        with nogil:
+            error = bhyve.vm_suspend(self.vm, vmm.VM_SUSPEND_RESET)
+
+        if error:
```
It seems libvmmapi simply returns errno from vmm ioctl, which usually are errno.h codes?
Perhaps we could also return the error string from errno?

sonicaj replied:
That is a good idea and would increase the understanding of the underlying issue without having to google the description. I couldn't find relevant error strings in the header file, I think I'll have to make a dictionary or something like that for them. Do you think that would make sense ?

william-gr replied:
I imagine you can use strerror()

sonicaj committed:
Reviews Addressed (3)

sonicaj committed:
Retrive Active VCPU count for VM

This commit introduces the ability for the end user to find active vcpu count for a specified VM. Many thanks to @araujobsd for his help in making this work and walking me through the internals of bhyve.
Following functionalites have been introduced:
1) get_stats function has been improved which now retrieves stats for all active vcpus of a vm by default and if only stats are required for a specified vm, that can be specified as well.
2) active_vcpus has been introduced which will return a count of active vcpu for the specified vm.

araujobsd commented on bhyve.pyx:
```diff
@@ -1,4 +1,139 @@
+
+        return True
+
+    def get_stats(self, vcpu=None):
```
The purpose of get_stats function is mostly to be used by bhyve developers as a way to see what is going on at guest side.

araujobsd commented on bhyve.pyx:
```diff
@@ -1,4 +1,139 @@
+        active_vcpus = self.active_vcpus()
+        if vcpu is not None and not -1 < vcpu < active_vcpus:
+            raise BhyveException(
+                Error.STATS_FETCH_FAILED, f'Please specify a vcpu number from 0-{active_vcpus}'
```
The error message here seems wrong, when we call this method we don't specify any vcpu, also vcpus should start from 1 to maximum 16 vcpus, although when you call libvmmapi, it will shows vcpu starting from 0 such like: 0, 1, 2, 3 for a guest with 4 vcpus.

araujobsd commented on bhyve.pyx:
```diff
@@ -1,4 +1,139 @@
+
+        if create_vm:
+            with nogil:
+                code = vmmapi.vm_create(vm_name)
```
from libvmmapi vm_create only creates the guest name space inside vmm device. I understand this module is under development, just to let you aware of it.

araujobsd commented on pxd/cpuset.pxd:
```diff
@@ -0,0 +1,4 @@
+cdef extern from 'sys/cpuset.h' nogil:
+    ctypedef unsigned long cpuset_t
+    extern bint CPU_EMPTY(cpuset_t *cpuset)
+    extern bint CPU_ISSET(size_t cpu_idx, cpuset_t	*cpuset)
```
Extra spaces on "cpuset_t *cpuset"

araujobsd commented on bhyve.pyx:
```diff
@@ -1,4 +1,139 @@
+cdef class VM:
+
+    cdef str vm_name
+    cdef vmmapi.vmctx* vm
```
C pointer must always be declared as ```type *variable```, here it doesn't create any issue, some people declare like ```type * variable``` others ```type* variable```, but on FreeBSD we always use "type *variable".

araujobsd commented on bhyve.pyx:
```diff
@@ -1,4 +1,139 @@
+    def __cinit__(self, str vm_name_p, bint create_vm=0):
+        self.vm_name = vm_name_p
+        vm_name_bytes = vm_name_p.encode()
+        cdef char* vm_name = vm_name_bytes
```
Here is the same with ```char* vm_name```.

araujobsd reviewed (approved):
Very good work with the active_vcpus(), I made couple tests and this function works great.

```
>>> c.active_vcpus()
8
>>> a.active_vcpus()
2
```

sonicaj committed:
Readme added
